dali/kernels/shape.h

```diff
@@ -0,0 +1,430 @@
+  const_reverse_iterator rbegin() const noexcept { return shape.rbegin(); }
+  const_reverse_iterator rend() const noexcept { return shape.rend(); }
+  const_reverse_iterator crbegin() const noexcept { return shape.crbegin(); }
+  const_reverse_iterator crend() const noexcept { return shape.crend(); }
+
+  template <int other_ndim>
+  TensorShape<other_ndim> first() {
+    static_assert(0 <= other_ndim && other_ndim <= ndim,
+                  "Number of elements in subshape must be between 0 and size()");
+    TensorShape<other_ndim> result;
+    for (int i = 0; i < other_ndim; i++) {
+      result[i] = (*this)[i];
+    }
+    return result;
+  }
+
+  template <int other_ndim>
+  TensorShape<other_ndim> last() {
+    static_assert(0 <= other_ndim && other_ndim <= ndim,
+                  "Number of elements in subshape must be between 0 and size()");
+    TensorShape<other_ndim> result;
+    int start_offset = ndim - other_ndim;
+    for (int i = 0; i < other_ndim; i++) {
+      result[i] = (*this)[start_offset + i];
+    }
+    return result;
+  }
+
+  std::array<int64_t, ndim> shape;
+  constexpr int size() const { return ndim; }
+};
+
+template <int left_ndim, int right_ndim>
+bool operator==(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  if (left.size() != right.size()) {
+    return false;
+  }
+  int size = left.size();
+  for (int i = 0; i < size; i++) {
+    if (left[i] != right[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator!=(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  return !(left == right);
+}
+
+template <int left_ndim, int right_ndim>
+typename std::enable_if<left_ndim != DynamicDimensions && right_ndim != DynamicDimensions, 
+                        TensorShape<left_ndim + right_ndim>>::type 
+    operator+(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  TensorShape<left_ndim + right_ndim> result;
+  for (int i = 0; i < left_ndim; i++) {
+    result[i] = left[i];
+  }
+  for (int i = 0; i < right_ndim; i++) {
+    result[left_ndim + i] = right[i];
+  }
+  return result;
+}
+
+template <int sample_ndim>
+typename std::enable_if<sample_ndim != DynamicDimensions, std::vector<int64_t>>::type
+flatten_shapes(const std::vector<TensorShape<sample_ndim>> &shapes) {
+  std::vector<int64_t> result;
+  result.resize(sample_ndim * shapes.size());
+  for (int sample = 0; sample < shapes.size(); sample++) {
+    for (int axis = 0; axis < sample_ndim; axis++) {
+      result[sample * sample_ndim + axis] = shapes[sample][axis];
+    }
+  }
+  return result;
+}
+
+template <typename T>
+typename std::enable_if<std::is_same<T, TensorShape<DynamicDimensions>>::value ||
+                            std::is_same<T, std::vector<int64_t>>::value,
+                        std::vector<int64_t>>::type
+flatten_shapes(const std::vector<T> &shapes, int uniform_sample_ndim) {
```

Comment:
Why not just take the first shape in the vector and use it to get `uniform_sample_ndim`?

Reply:
My idea was to make it clear, that shapes should have the same dimensions. Now you have to ensure this while calling those functions/constructor of TensorListShape.

I already encountered the example of taking the shape of the first sample in our code, and it was a bit broken, so I decided that this is safer.